src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatDefaultPayload.cs

```diff
@@ -42,6 +100,84 @@ public static void PopulateDefaultPayload(IEnumerable<string> disabledFields, IH
             }
         }
 
+        /// <summary>
+        /// Gathers operational data from an Azure VM, if the IMDS endpoint is present. If not, and the SDK is
+        /// not running on an Azure VM with the Azure Instance Metadata Service running, the corresponding
+        /// fields will not get set into the heartbeat payload.
+        /// </summary>
+        private static async Task AddAzureVmDetail(HeartbeatProvider heartbeatManager, IEnumerable<string> enabledFields)
+        {
+            if (!HeartbeatDefaultPayload.hasAzureVmMetadataBeenGathered)
+            {
+                // only ever do this once when the SDK gets initialized
+                HeartbeatDefaultPayload.hasAzureVmMetadataBeenGathered = true;
+
+                try
+                {
+                    var allFields = await GetAzureInstanceMetadataFields(baseImdsUrl, imdsApiVersion, imdsTextFormat)
+                                    .ConfigureAwait(false);
+
+                    var enabledImdsFields = enabledFields.Intersect(allFields);
+                    foreach (string field in enabledImdsFields)
+                    {
+                        heartbeatManager.AddHeartbeatProperty(
+                            propertyName: field,
+                            propertyValue: await GetAzureInstanceMetadaValue(baseImdsUrl, field, imdsApiVersion, imdsTextFormat)
```

Comment:
if this AzureMetaDataService has a temp. issue, we need to retry. (500 code) It might be better to handle exceptions/etc inside this helper method itself rather than an overall try catch.

Reply:
Ok. Should we do a retry on this though since the Azure instance service should be stable and available while we run our app. If it isn't, it could be that it would never be stable and would be a waste of our resources to continue to try... I was going for a fire-once-and-forget pattern here, failures or no.